src/chainstate/stacks/boot/pox-3.clar

```diff
@@ -0,0 +1,1308 @@
+      ;;   before it can be included in the reward set, this must be committed!
+      (add-pox-partial-stacked pox-addr first-increase-cycle cycle-count increase-by)
+
+      ;; stacking-state is unchanged, so no need to update
+
+      ;; return the lock-up information, so the node can actually carry out the lock.
+      (ok { stacker: stacker, total-locked: new-total-locked}))))
+
+;; As a delegator, extend an active stacking lock, issuing a "partial commitment" for the
+;;   extended-to cycles.
+;; *New in Stacks 2.1*
+;; This method extends `stacker`'s current lockup for an additional `extend-count`
+;;    and partially commits those new cycles to `pox-addr`
+(define-public (delegate-stack-extend
+                    (stacker principal)
+                    (pox-addr { version: (buff 1), hashbytes: (buff 32) })
+                    (extend-count uint))
+    (let ((stacker-info (stx-account stacker))
+          ;; to extend, there must already be an entry in the stacking-state
+          (stacker-state (unwrap! (get-stacker-info stacker) (err ERR_STACK_EXTEND_NOT_LOCKED)))
+          (amount-ustx (get locked stacker-info))
+          (unlock-height (get unlock-height stacker-info))
+          (unlock-in-cycle (burn-height-to-reward-cycle unlock-height))
+          ;; if the account unlocks *during* this cycle (should only occur during testing),
+          ;; set first-extend-cycle to the next cycle.
+          (cur-cycle (current-pox-reward-cycle))
+          (first-extend-cycle (if (> (+ cur-cycle u1) unlock-in-cycle)
+                                     (+ cur-cycle u1) unlock-in-cycle))
+          ;; new first cycle should be max(cur-cycle, stacker-state.first-reward-cycle)
+          (cur-first-reward-cycle (get first-reward-cycle stacker-state))
+          (first-reward-cycle (if (> cur-cycle cur-first-reward-cycle) cur-cycle cur-first-reward-cycle)))
+
+     ;; must be called with positive extend-count
+     (asserts! (>= extend-count u1)
+               (err ERR_STACKING_INVALID_LOCK_PERIOD))
+
+     (let ((last-extend-cycle  (- (+ first-extend-cycle extend-count) u1))
+           (lock-period (+ u1 (- last-extend-cycle first-reward-cycle)))
+           (new-unlock-ht (reward-cycle-to-burn-height (+ u1 last-extend-cycle))))
+
+      ;; first cycle must be after the current cycle
+      (asserts! (> first-extend-cycle cur-cycle) (err ERR_STACKING_INVALID_LOCK_PERIOD))
+      ;; lock period must be positive
+      (asserts! (> lock-period u0) (err ERR_STACKING_INVALID_LOCK_PERIOD))
+
+      ;; must be called directly by the tx-sender or by an allowed contract-caller
+      (asserts! (check-caller-allowed)
+        (err ERR_STACKING_PERMISSION_DENIED))
+
+      ;; stacker must be delegated to tx-sender
+      (asserts! (is-eq (unwrap! (get delegated-to stacker-state)
+                                (err ERR_STACKING_NOT_DELEGATED))
+                       tx-sender)
+                (err ERR_STACKING_PERMISSION_DENIED))
```

Comment:
You should also add a `(asserts! (is-eq 0 (len (get reward-set-indexes ..)` check here (just as a sanity check).